nemo/collections/asr/metrics.py

```diff
@@ -97,3 +98,26 @@ def classification_accuracy(
             results.append(correct_k)
 
     return results
+
+
+def classification_confusion_matrix(logits: torch.Tensor, targets: torch.Tensor,) -> torch.Tensor:
+    """
+    Computes the confusion matric provided with 
+    un-normalized logits of a model and ground truth targets.
+
+    Args:
+        logits: Un-normalized logits of a model. Softmax will be
+            applied to these logits prior to computation of accuracy.
+        targets: Vector of integers which represent indices of class
+            labels.
+    Returns:
+
+        A tensor (n_classes, n_classes) of confusion matrix.
+    """
+
+    with torch.no_grad():
+        _, predictions = logits.topk(1, dim=1, largest=True, sorted=True)
+        predictions = predictions.t().squeeze()
+        targets = targets.t().squeeze()
+
+    return sklearn.metrics.confusion_matrix(targets, predictions, labels=range(logits.shape[1]))
```

Comment:
This makes the assumption that input is a 2D tensor of class labels (with only 2 classes). Either enforce this assumption with a check at the top before computing the predictions, or expose additional parameters to enable multiclass confusion matrix computation. (micro, macro schemes should be supported in such a case if we allow more than 2 classes).

Also add tests for this metric.

Reply:
Do we allow multidimensional batch? like (2,128,3), where (2,128) is the batch shape?

Reply:
If we don't allow the above condition. We don't need to assert before computing the predictions I guess. The code would automatically enable multiclass confusion matrix. What do you mean micro and macro schemes here for the confusion matrix?

Reply:
sklearn classification_matrix doesnt support multidim batch, so neither should we. About micro/macro, thats for the f1 score calculation, not for confusion matrix. My mistake. However if we have more than 2 classes, your TP/TN/FP/FN calculation later on fails, so either force this code to accept only binary matrices, or perform a check below to ensure you can properly calculate those values.

Since you calculate those values just to compute Precision/Recall/F1, you might as well return the argmaxed values, and use the sklearn metrics directly to enable multiclass computation of those scores (which then brings up micro/macro multiclass score computation).

Look at https://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html (specifically the `average='binary'` part).